docs/guides/single-fetch.md

```diff
@@ -0,0 +1,155 @@
+
+- If you are already returning `json`/`defer` responses from your loader/action functions, they'll work the same after Single Fetch is enabled
+- If you are returning naked objects, they will no longer be automatically converted to JSON, and will be streamed down as-is
+  - This means that `return { date: new Date() }` from your loader will result in `useLoaderData().date` being a `Date` instance (instead of a`string` as it is today).
+  - If you wish to maintain current behavior, just wrap any existing naked object returns in `json`
+
+This also means that you no longer need to use the `defer` utility to send `Promise` instances over the wire! You can include a `Promise` anywhere in the native object and pick it up on `useLoaderData().whatever`. You can also nest `Promise`'s if needed!
+
+### React Rendering APIs
+
+In order to maintain consistency between document and data requests, `turbo-stream` is also used as the format for sending down data in initial document requests. This means that once opted-into Single Fetch, your application can no longer use [`renderToString`][rendertostring] and must use a React streaming renderer API such as [`renderToPipeableStream`][rendertopipeablestream] or [`renderToReadableStream`][rendertoreadablestream]) in [`entry.server.tsx`][entry-server].
+
+This does not mean you _have_ to stream down your HTTP response, you can still send the full document at once by leveraging the `onAllReady` option in `renderToPipeableStream`, or the `allReady` promise in `renderToReadableStream`.
+
+On the client side, this also means that your need to wrap your client-side [`hydrateRoot`][hydrateroot] call in a [`startTransition`][starttransition] call because the streamed data will be coming down wrapped in a `Suspense` boundary.
+
+### Streaming Timeout
+
+Previously, Remix has a concept of an `ABORT_TIMEOUT` built-into the default [`entry.server.tsx`][entry-server] files which would terminate the React renderer, but it didn't do anything in particular to clean up any pending deferred promises.
+
+Now that Remix is streaming internally, we can cancel the `turbo-stream` processing and automatically reject any pending promises and stream up those errors to the client. By default, this happens after 4950ms - a value that was chosen to be just under the current 5000ms `ABORT_DELAY` in most entry.server.tsx files - since we need to cancel the promises and let the rejections stream up through the React renderer prior to aborting the React sid eof things.
+
+You can control this by exporting a `streamTimeout` numeric value from your `entry.server.tsx` and Remix will use that as the number of milliseconds after which to reject any outstanding Promises from `loader`/`action`'s. It's recommended to decouple this value from the timeout in which you abort the React renderer - and you should always set the React timeout to a higher value so it has time to stream down the underlying rejections from your `streamTimeout`.
```

Comment:
What if instead of `export const streamTimeout = 10_000` we could do:
```ts
export const config: RemixServerConfig = {
  streamTimeout: 10_000
}
```
Then future options could be added to `RemixServerConfig` without adding more exports

Reply:
Not sure if this is the right place to discuss this, but since you propose a code change, I'll jump in: it would be beneficial to let the code make the decision suggested in the docs. That way we remove a trip-point. Instead of just having docs specifying that you must ensure stream timeout to be lesser than the `ABORT_DELAY` configuration, set it up by calling a function that takes care of this. Something like:

```ts
// `lib.ts`, or maybe in `vite.config.ts`

// define this somewhere else and export it
import { ABORT_DELAY } from "config.ts"

export const getProperTimeout = () => {
  const timeDifferenceInMs = 10
  return ABORT_DELAY - timeDifferenceInMs
}
```

Reply:
@sergiodxa - the exported object is definitely nice for typings, but I don't foresee us adding _too_ many more exports and I have a hunch that you'll have more first-hand control over this in the future when we move towards an RSC-enabled landscape. So I don't know if it makes sense to introduce an object export which may never grow beyond 1-2 fields.

@fernandocanizo - I don't personally like the idea of moving these configs anywhere farther from entry-server since they're specifically runtime (and server-only) decisions - whereas fields in the configs are mostly (if not exclusively) build-time decisions. I think no matter what, there will be some developer-education needed which we'll do via docs or even comments in the template source code.  In your proposed API if `getProperTimeout` is not exported by the developer, then the same potential issue remains as if they didn't set them correctly.  I think the vast majority of current Remix devs never touch `ABORT_DELAY` currently and the ones who do are experienced enough to know the implications.  I think the same will be true in the future with `streamTimeout`.